### ISSUE TITLE
Fixture loading makes before(:all) unpredictable

### ISSUE THREAD

**Issue by virolea** [rspec/rspec-rails#2816]
## What Ruby, Rails and RSpec versions are you using?

Ruby version: 3.3.0
Rails version: 8.0
RSpec version: 3.13

## Observed behaviour

I am using both factories (with FactoryBot) and fixtures in my application. For some models, I have both factories and fixtures. 

To speed some test groups, I am starting to implement some `before(:all)` hooks in my specs to create test data only once. However, some specs started to fail after moving to this new setup. This only happens when I am running a particular spec, not when running the whole suite. 

Here's a simple example that features the problem (a corresponding users.yml fixture file should exist)

```ruby
require 'rails_helper'

RSpec.describe User, type: :model do
  before(:all) do
    @user = FactoryBot.create(:user)
  end

  after(:all) do
    @user.destroy
  end

  # fails when run separately
  it "exists" do
    expect(User.exists?(@user.id)).to eq true
  end
end

```

## Expected behaviour

The specs should not fail 

## Can you provide an example reproduction?

From my research, it looks like the fixtures loading happens after the before(:all) setup. Since fixture loading destroys previously set data for a given object, if an instance was created previous to the fixtures loading, it will be wiped out. 

Here's an example repo that features the problem https://github.com/virolea/test-app-rspec

```
bundle install 
bundle exec rspec
```

Ideally, fixture loading should happen prior to user-defined before(:all) hooks

**Comment by pirj** [comment]
This is as expected, no?
I certainly understand your desire to have clean fixtures on each example, and some (potentially heavy) factories created just once, but this is not trivial to achieve not to say is available out of the box.
Have a look at what test-prof has to offer, but I can’t tell off the top of my head if its before_all/let_it_be would work with fixtures.

**Comment by virolea** [comment]
This is as expected, no?
I certainly understand your desire to have clean fixtures on each example, and some (potentially heavy) factories created just once, but this is not trivial to achieve not to say is available out of the box.
Have a look at what test-prof has to offer, but I can’t tell off the top of my head if its before_all/let_it_be would work with fixtures.

Hey @pirj thanks for getting back to me so quickly. 

> This is as expected, no?

From my point of view it was not. I would assume the test setup would be done (and hence fixtures cleaned then loaded) before any test starts. Having my data set in the before(:all) hook wiped out before the test was something I was not expecting and lead to quite a bit of research. 

I am not familiar with the code so i don't know the implications behind this design. This is not a blocker as the tests pass at the suite level. Not ideal though.

**Comment by pirj** [comment]
Can you please provide a minimal repro app/script?
Are transactions turned on?
Is fixture cleanup done by rolling back the test transaction?
Do factories run before the transaction start in before(:all)?
If not, why?
Is databaseeaner used?

**Comment by pirj** [comment]
Sorry for the confusion. Let me rephrase.

What runs earlier, the code that is inside ‘before(:all)’s block, or the very first transaction?

You may get some insights in log/test.log, as usually SQL statements are logged.

**Comment by virolea** [comment]
The code in the before(:all) block runs first.

However it runs prior to Fixture loading, as shown in the logs below: 

![CleanShot 2024-11-28 at 10 20 39@2x](https://github.com/user-attachments/assets/23e49621-3181-4336-b47a-11b21a2bbad2)

**Comment by pirj** [comment]
Can you pinpoint this deletion to some Rails code?

**Comment by virolea** [comment]
The deletion comes from Fixtures loading. From [the docs](https://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html#class-ActiveRecord::FixtureSet-label-Using+Fixtures+in+Test+Cases)

> The testing environment will automatically load all the fixtures into the database before each test. To ensure consistent data, **the environment deletes the fixtures before running the load.**

RSpec Rails loads fixtures in 

https://github.com/rspec/rspec-rails/blob/609bb83a76c11c67076ff2619fd5f47edd6325ce/lib/rspec/rails/fixture_support.rb#L10

`TestFixtures` setups fixtures in a `before_setup` hook that is called via `RSpec::Rails::MinitestLifecycleAdapter`

https://github.com/rspec/rspec-rails/blob/609bb83a76c11c67076ff2619fd5f47edd6325ce/lib/rspec/rails/adapters.rb#L66-L78

This is the code from Rails itself that setups fixtures: 

```ruby
module ActiveRecord
  module TestFixtures
    extend ActiveSupport::Concern

    def before_setup # :nodoc:
      setup_fixtures
      super
    end
```

and if you follow deep enough the `setup_fixtures` call, [you end up to the code](https://github.com/rails/rails/blob/8b544b2689f97f57c97bee59f45920bd99bda047/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb#L488) that deletes tables that have fixtures counterpart.

**Comment by JonRowe** [comment]
:wave: Sorry you're experiencing this difficulty, I'm not sure how solvable this is... 

A bit of context, a`before(:context)` hook will always execute before any `before(:example)` (or `around`) hook, that is expected (and [documented](https://rspec.info/features/3-13/rspec-core/hooks/before-and-after-hooks/)) behaviour, and an unavoidable part of their design. So its impossible for something to be executed before all examples, that is also after an individual before hook.

And as you have seen we are not implementing fixtures ourselves, just integrating the Rails API in a similar place to how it was built. This means that fixtures are called before every spec, that is by the design of the fixture setup.

So I'm actually surprised your code works at all, I would expect it to always be flaky depending when a fixture runs.

I think the solution here is to migrate to factories entirely if you want this sort of setup.

**Comment by jakeonfire** [comment]
fwiw we have a similar requirements, but we are _not_ setting `config.global_fixtures = :all` but instead declaring `fixtures :all` in spec files where we need them. then, in files where we don't do so, setting up factories in a `before(:all)` works just fine. it looks like using fixtures (globally or within a top-level context) _and_ setting up factories in a `before(:all)` is not possible.
